README.md
# i18n-for-browser

[![NPM version][npm]][npm-url]
[![Dependencies status][deps]][deps-url]
[![Build status][build]][build-url]
[![Coverage status][coverage]][coverage-url]
[![Dependabot badge][dependabot]][dependabot-url]
[![Documentation badge][documentation]][documentation-url]

[npm]: https://img.shields.io/npm/v/i18n-for-browser.svg
[npm-url]: https://npmjs.com/package/i18n-for-browser

[deps]: https://david-dm.org/TrigenSoftware/i18n-for-browser.svg
[deps-url]: https://david-dm.org/TrigenSoftware/i18n-for-browser

[build]: http://img.shields.io/travis/com/TrigenSoftware/i18n-for-browser/master.svg
[build-url]: https://travis-ci.com/TrigenSoftware/i18n-for-browser

[coverage]: https://img.shields.io/coveralls/TrigenSoftware/i18n-for-browser.svg
[coverage-url]: https://coveralls.io/r/TrigenSoftware/i18n-for-browser

[dependabot]: https://api.dependabot.com/badges/status?host=github&repo=TrigenSoftware/i18n-for-browser
[dependabot-url]: https://dependabot.com/

[documentation]: https://img.shields.io/badge/API-Documentation-2b7489.svg
[documentation-url]: https://trigensoftware.github.io/i18n-for-browser

Modern translation module for web.

## Install

```sh
npm i i18n-for-browser
# or
yarn add i18n-for-browser
```

## CDN

`i18n-for-browser` is also available on [unpkg.com](https://unpkg.com/i18n-for-browser) as UMD, which exposes global object `i18n`.

```html
<script src="https://unpkg.com/i18n-for-browser?main=umd"></script>
```

## API

Module exposes next API:

```js
export default globalConfig;
export {
    I18nConfigInstance,
    I18nPluralLocale,
    I18nLocale,
    I18nLangLocales,
    I18nLocales,
    I18nFallbacks,
    I18nUnknownPhraseListener,
    I18nProcessor,
    I18nConfig,
    I18nForkConfig,
    I18nParams,
    I18nPluralParams,
    pluralIntervalProcessor,
    mustacheProcessor,
    __,
    __mf,
    __n,
    __m
};
```

[Description of this methods you can find in Documentation.](https://trigensoftware.github.io/i18n-for-browser/index.html)

Shirt description:

### [i18n](https://trigensoftware.github.io/i18n-for-browser/modules/_src_methods_common_.html#globalconfig)

Global config. Instanse of [`Config`](https://trigensoftware.github.io/i18n-for-browser/classes/_src_config_.config.html).

<details>
    <summary>Usage example</summary>

```js
import i18n, {
    pluralIntervalProcessor,
    __,
    __n
} from 'i18n-for-browser';

/**
 * Set global config.
 */
i18n.configure({
    /**
     * Store of translations.
     */
    locales: {
        'en': {
            /**
             * Simple translation example.
             */
            'cat': 'cat',
            /**
             * Plutal translation example.
             */
            '%s cats': {
                'one': '%s cat',
                'other': '%s cats'
            },
            /* ... */
        },
        'ru': {
            /**
             * Пример простого перевода.
             */
            'cat': 'кошка',
            /**
             * Пример перевода множественного числа.
             */
            '%s cats': {
                'one': '%s кошка',
                'few': '%s кошки',
                'many': '%s кошек',
                'other': '%s кошка'
            },
            /* ... */
        }
    },
    /**
	 * Cookie name to store locale.
	 */
    cookieName: 'yourcookiename'
});

console.log(__('cat')); // Uses global config.

/**
 * Create config fork with some overrides. 
 */
const i18nFork = i18n.fork({
    /**
	 * List of post processors.
	 */
    processors: [pluralIntervalProcessor]
});
/**
 * Bind new config to method.
 */
const __pi = i18nFork.bind(__n);

/**
 * Now you able to use plural intervals.
 */
console.log(
    __pi('[0] no dog|[2,5] some dogs|[6,11] many dogs|[12,36] dozens of dogs|a horde of %s dogs|[100,] too many dogs', 3) // Uses bound config.
);
```

</details>

### [__()](https://trigensoftware.github.io/i18n-for-browser/modules/_src_methods____.html#__)

Translates a single phrase and adds it to locales if unknown.

<details>
    <summary>Usage example</summary>

```js
/**
 * Basic usage 
 */
__('cat')
/**
 * As template string
 */
__`cat`
/**
 * Supports sprintf formatting
 */
__('%d cats', 3)
/**
 * Sprintf formatting with template string
 */
__`${3} cats`
/**
 * Sprintf formatting with few arguments
 */
__('%d cats with %s', 3, 'long tails')
/**
 * Mustache templates are supported with `mustacheProcessor`
 */
__('Hello {{name}}', { name: 'Marcus' })
/**
 * First argument as object with specified locale
 */
__({ phrase: 'Hello', locale: 'ru' })
```

</details>

### [__mf()](https://trigensoftware.github.io/i18n-for-browser/modules/_src_methods___mf_.html#__mf)

Supports the advanced MessageFormat as provided by excellent messageformat module. You should definetly head over to messageformat.github.io for a guide to MessageFormat. `i18n-for-browser` takes care of `new MessageFormat('en').compile(msg);` with the current msg loaded from it's json files and cache that complied fn in memory. So in short you might use it similar to `__()` plus extra object to accomblish MessageFormat's formating.

<details>
    <summary>Usage example</summary>

```js
/**
 * Basic usage, also works as raw `__` method
 */
__mf('cat')
/**
 * Basic replacement
 */
__mf('Hello {name}', { name: 'Marcus' })
/**
 * Also work with sprintf formatting
 */
__mf('Hello {name}, how was your %s?', 'test', { name: 'Marcus' })
```

</details>

### [__n()](https://trigensoftware.github.io/i18n-for-browser/modules/_src_methods___n_.html#__n)

Plurals translation of a single phrase. Singular and plural forms will get added to locales if unknown. Returns translated parsed and substituted string based on `count` parameter.

<details>
    <summary>Usage example</summary>

```js
/**
 * Basic usage
 */
__('%s cats', 2)
/**
 * As template string
 */
__`${3} cats`
/**
 * Can work without translation in config
 */
__('%d dog', '%d dogs', 3)
/**
 * First argument as object with specified locale
 */
__n({ singular: '%s cat', plural: '%s cats', locale: 'nl', count: 3 })
```

</details>

### [__m()](https://trigensoftware.github.io/i18n-for-browser/modules/_src_methods___m_.html#__m)

Returns a map of translations for a given phrase in each language.

<details>
    <summary>Usage example</summary>

```js
/**
 * Basic usage
 */
__m(__, 'Hello')
```

</details>

## Express middleware helper

To provide translations to client from your express app you can use this helper.

```js
import i18nExpressHelper from 'i18n-for-browser/lib/middleware';
// or 
const i18nExpressHelper = require('i18n-for-browser/lib/middleware');
// ...
// Before this `i18n` should already initialized.
app.use(i18nExpressHelper(i18nNodeConfig));
// ...
```

```html
<!DOCTYPE html>
<html>
    <head>
        <title>views/layout.ejs</title>
        <%- initI18nForBrowser() %>
        <script src="js-file-with-imported-i18n-for-browser.js"></script>
    </head>
    <body>
        <%- body %>
    </body>
</html>
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

<!--

DO NOT TOUCH. SAVE IT ON TOP.

## [semver] - date
### Added
- ...

### Changed
- ...

### Fixed
- ...

### Removed
- ...

-->

## [3.0.0] - 2020-10-02
### Changed
- Types names.

## [2.1.0] - 2020-05-22
### Changed
- Dependencies update.

## [2.0.0] - 2020-03-18
### Added
- Babel module.

### Removed
- Drop Node 8 support.

## [1.1.0] - 2019-06-09
### Added
- `trigen-scripts` dev tool.

### Changed
- Dependencies update.

## [1.0.1] - 2018-11-20
### Fixed
- Template literal tag typings fix.

## [1.0.0] - 2018-11-15
### Added
- `size-limit`
- [Greenkeeper](https://greenkeeper.io/)
- `__m` instead of `__l` and `__h`.
- `unknownPhraseListener`

### Changed
- Rewritten to TypeScript.
- Improvements for tree-shaking: `__`, `__n`, `__mf` and `__m` now are importable functions; Mustaches and plural intervals moved to pluggable postprocessor-functions.  
- Tests and code coverage with `jest`.
- `cookie` option -> `cookieName`.
- `objectNotation` option now is `boolean`, `.` is constant separator.

### Fixed
- `parsePluralInterval`

### Removed
- `__l` and `__h`.
- `localeChangeListener`
- Getting `count` for plurals from named values.
- `globalize` option.

## [0.9.7] - 2017-11-08
### Added
- Code coverage with `nyc`.
- Integration with [Coveralls](https://coveralls.io).

### Changed
- `"browser"` property in `package.json` renamed to `"umd"` due to webpack uses `"browser"` first.
- Link to UMD bundle on [unpkg.com](https://unpkg.com) changed.

### Fixed
- UMD bundle fixed.


src/lib.js
/**
 * @author      Created by Dan Green <danon0404@gmail.com> on 2016-01-05.
 * @link        https://github.com/TrigenSoftware/i18n-for-browser
 * @license     http://opensource.org/licenses/MIT
 *
 * @version     0.5.3
 */

import { vsprintf }  from 'sprintf-js';
import ParseInterval from 'math-interval-parser';
import MakePlural    from 'make-plural/make-plural';
import Plurals       from 'make-plural/data/plurals.json';
import MessageFormat from 'messageformat';
import Mustache      from 'mustache';
import Url           from 'url';

MakePlural.load(Plurals);

var localeChangeListener = () => {},
	defaultLocale  = 'en',
	locales        = {},  
	fallbacks      = {}, 
	cookiename     = null, 
	objectNotation = false,

	MessageFormatInstanceForLocale = {},
	PluralsForLocale = {};

// Silent configure if I18N object is exist.
if (typeof global.I18N == "object") {
	configure(I18N);
}

/**
 * Configure i18n with given options.
 * 
 * @param  {Object} options
 */
export function configure(options) {

	// read languages locales
	if (typeof options.locales == 'object') {
		locales = options.locales;
	}

	// sets a custom cookie name to parse locale settings from
	if (typeof options.cookie == 'string') { 
		cookiename = options.cookie;
	}

	// setting defaultLocale
	if (typeof options.defaultLocale == 'string') {
		defaultLocale = options.defaultLocale;
	}

	// enable object notation?
	if (typeof options.objectNotation != 'undefined') { 
		objectNotation = options.objectNotation;
	}

	if (objectNotation === true) {
		objectNotation = '.';
	}

	// read language fallback map
	if (typeof options.fallbacks == 'object') {
		fallbacks =  options.fallbacks;
	}

	// globalize
	if (options.globalize === true) {
		globalize();
	}

	defaultLocale = getCookie(cookiename) || defaultLocale;

	// get default locale from url
	var typeofQueryParameter = typeof options.queryParameter;
	if (typeofQueryParameter != 'undefined' && typeof location != "undefined") {

		var queryParameter = typeofQueryParameter == 'string' 
			? options.queryParameter
			: 'locale'; 

		var localeFromQuery = Url(location.href).query['queryParameter'];

		if (typeof localeFromQuery == "string") {
			defaultLocale = localeFromQuery;
		}
	}
}

/**
 * Inject `__`, `__n` and other functions to global scope.
 * 
 */
export function globalize() {
	global.__   = applyAPItoObject(__);
	global.__n  = applyAPItoObject(__n);
	global.__l  = applyAPItoObject(__l);
	global.__h  = applyAPItoObject(__h);
	global.__mf = applyAPItoObject(__mf);
}


/**
 * Translates a single phrase and adds it to locales if unknown. 
 * Returns translated parsed and substituted string.
 * 
 * @param  {String}    phrase
 * @param  {...Object} params
 * @return {String}    translate
 */
export function __(_phrase, ...params) {

	var phrase = preProcess(_phrase),
		translated, namedValues;

	// Accept an object with named values as the last parameter
	if (typeof params[params.length - 1] == "object") {
		namedValues = params.pop();
	}

	// called like __({phrase: "Hello", locale: "en"})
	if (typeof phrase === 'object') {

		if (typeof phrase.locale === 'string' && typeof phrase.phrase === 'string') {
			translated = translate(phrase.locale, phrase.phrase);
		}
	}
	// called like __("Hello")
	else {
		// get translated message with locale
		translated = translate(defaultLocale, phrase);
	}

    // postprocess to get compatible to plurals
    if (typeof translated === 'object' && typeof translated.one != 'undefined') {
    	translated = translated.one;
    }

    // in case there is no 'one' but an 'other' rule
    if (typeof translated === 'object' && typeof translated.other != 'undefined') {
    	translated = translated.other;
    }

    return postProcess(translated, namedValues, params);
}

/**
 * Supports the advanced MessageFormat as provided by excellent messageformat module. 
 * You should definetly head over to messageformat.github.io for a guide to MessageFormat. 
 * i18n takes care of new MessageFormat('en').compile(msg); 
 * with the current msg loaded from it's json files and cache that complied fn in memory. 
 * So in short you might use it similar to __() plus extra object to accomblish MessageFormat's formating.
 * 
 * @param  {String}    phrase
 * @param  {...Object} params
 * @return {String}    translate
 */
export function __mf(phrase, ...params) {

	var targetLocale = defaultLocale,
		translated, namedValues,
		mf, f;

	// Accept an object with named values as the last parameter
	if (typeof params[params.length - 1] == "object") {
		namedValues = params.pop();
	}

	// called like __mf({phrase: "Hello", locale: "en"})
	if (typeof phrase === 'object') {

		if (typeof phrase.locale === 'string' && typeof phrase.phrase === 'string') {
			targetLocale = phrase.locale;
			phrase       = phrase.phrase;
		}
	}
	// else called like __mf("Hello")

	translated = translate(targetLocale, phrase);
    // --- end get translate

    // now head over to MessageFormat
    // and try to cache instance
    if (MessageFormatInstanceForLocale[targetLocale]) {
    	mf = MessageFormatInstanceForLocale[targetLocale];
    } else {
    	mf = new MessageFormat(targetLocale);
    	mf.compiledFunctions = {};
    	MessageFormatInstanceForLocale[targetLocale] = mf;
    }

    // let's try to cache that function
    if (mf.compiledFunctions[translated]) {
    	f = mf.compiledFunctions[translated];
    } else {
    	f = mf.compile(translated);
    	mf.compiledFunctions[translated] = f;
    }

    return postProcess(f(namedValues), namedValues, params);
}

/**
 * Plurals translation of a single phrase. 
 * Singular and plural forms will get added to locales if unknown. 
 * Returns translated parsed and substituted string based on `count` parameter.
 * 
 * @param  {String}    singular
 * @param  {String}    plural  
 * @param  {Number}    count   
 * @param  {...Object} params  
 * @return {String}    translate
 */
export function __n(_singular, plural, count, ...params) {

	var singular = preProcess(_singular),
		translated, namedValues, targetLocale;

	// Accept an object with named values as the last parameter
	if (typeof params[params.length - 1] == "object") {
		namedValues = params.pop();
	}

	// called like __n({singular: "%s cat", plural: "%s cats", locale: "en"}, 3)
	if (typeof singular === 'object') {

		if (typeof singular.locale === 'string' && typeof singular.singular === 'string' && typeof singular.plural === 'string') {
			translated = translate(targetLocale = singular.locale, singular.singular, singular.plural);
		}

		params.unshift(count);

		// some template engines pass all values as strings -> so we try to convert them to numbers
		if (typeof plural === 'number' || parseInt(plural, 10) + '' === plural) {
			count = plural;
		}

		// called like __n({singular: "%s cat", plural: "%s cats", locale: "en", count: 3})
		if (typeof singular.count === 'number' || typeof singular.count === 'string') {
			count = singular.count;
			params.unshift(plural);
		}
	}
	else {
		// called like	__n('cat', 3)
		if (typeof plural === 'number' || parseInt(plural, 10) + '' === plural) {
			count = plural;

	        // we add same string as default
	        // which efectivly copies the key to the plural.value
	        // this is for initialization of new empty translations
	        plural = singular;

			params.unshift(count);
			params.unshift(plural);
		}
		// called like __n('%s cat', '%s cats', 3)
		// get translated message with locale from scope (deprecated) or object
		translated = translate(targetLocale = defaultLocale, singular, plural);
	}

	if (count === null) {
		count = namedValues.count;
	}

    // enforce number
    count = parseInt(count, 10);

    // find the correct plural rule for given locale
    if (typeof translated === 'object') {

    	var p;

		// create a new Plural for locale
		// and try to cache instance
		if (PluralsForLocale[targetLocale]) {
			p = PluralsForLocale[targetLocale];
		} else {
			p = new MakePlural(targetLocale);
			PluralsForLocale[targetLocale] = p;
		}

		// fallback to 'other' on case of missing translations
		translated = translated[p(count)] || translated.other;
    }

    return postProcess(translated, namedValues, params, count);
}

/**
 * Returns a list of translations for a given phrase in each language.
 * 
 * @param  {String} phrase
 * @return {Array<String>}
 */
export function __l(phrase) {
    return getLocales().sort().map( locale => __({ 
    	phrase, locale 
    }));
}

/**
 * Returns a hashed list of translations for a given phrase in each language.
 * 
 * @param  {String} phrase
 * @return {Array<Object>}
 */
export function __h(phrase) {
    return getLocales().sort().map( locale => ({ 
    	[locale] : __({ 
    		phrase, locale 
    	})
    }));
}

/**
 * Set function to call when locale was change.
 * 
 * @param  {Function} listener
 */
export function onLocaleChange(listener) {
	localeChangeListener = listener;
}

/**
 * Set current locale.
 * 
 * @param  {String} locale
 * @return {String} locale
 */
export function setLocale(locale) {

	var fallback = fallbacks[locale];

	if (typeof locales[locale] != "object" && typeof fallback == "string") {
		locale = fallback;
	}

	// called like setLocale('en')
	if (typeof locales[locale] == "object") {

		defaultLocale = locale;

		if (cookiename !== null) {
			setCookie(cookiename, defaultLocale, {
			    expires: 3600 * 24 * 31 * 12 * 100,
			    path:    "/"
			});
		}

		localeChangeListener(defaultLocale);
	}

	return defaultLocale;
}

/**
 * Get current locale.
 * 
 * @return {String}
 */
export function getLocale() {
	return defaultLocale;
}

/**
 * Get array of available locales.
 * 
 * @return {Array<String>}
 */
export function getLocales() {
	return Object.keys(locales);
}

/**
 * Returns a whole catalog optionally based on given locale.
 * 
 * @param  {String} locale
 * @return {Object|Array}
 */
export function getCatalog(locale) {

	if (typeof locale == "undefined") {
		return locales;
	}

	var fallback = fallbacks[locale];

	if (typeof locales[locale] != "object" && typeof fallback == "string") {
		locale = fallback;
	}

	var catalog = locales[locale];

	// called like setLocale('en')
	if (typeof catalog == "object") {
		return catalog;
	}

	return false;
}

/**
 * Add new translations.
 * 
 * @param {String} locale
 * @param {Object} catalog
 */
export function addLocale(locale, catalog) {
	locales[locale] = catalog;
}

/**
 * Remove translations.
 * 
 * @param  {String}
 */
export function removeLocale(locale) {
	delete locales[locale];
}


function applyAPItoObject(object) {

	object.onLocaleChange = onLocaleChange;
	object.setLocale      = setLocale;
	object.getLocale      = getLocale
	object.getLocales     = getLocales;
	object.getCatalog     = getCatalog;
	object.addLocale      = addLocale;
	object.removeLocale   = removeLocale;

	return object;
}

function setCookie(name, value, options) {

	if (typeof document == "undefined") {
		return;
	}

	options = options || {};

	var expires = options.expires;

	if (typeof expires == "number" && expires) {

  		var d = new Date();

    	d.setTime(d.getTime() + expires * 1000);
    	expires = options.expires = d;
  	}

	if (expires && expires.toUTCString) {
		options.expires = expires.toUTCString();
	}

	value = encodeURIComponent(value);

	var updatedCookie = name + "=" + value;

	for (var propName in options) {
  		
  		updatedCookie += "; " + propName;
    	
    	var propValue = options[propName];
    	
    	if (propValue !== true) {
    		updatedCookie += "=" + propValue;
      	}
    }

	document.cookie = updatedCookie;
}

function getCookie(name) {

	if (typeof document == "undefined" || name == null) {
		return false;
	}

	var matches = document.cookie.match(new RegExp(
		"(?:^|; )" + name.replace(/([\.$?*|{}\(\)\[\]\\\/\+^])/g, '\\$1') + "=([^;]*)"
	));

	return matches ? decodeURIComponent(matches[1]) : false;
}


function preProcess(text) {

	if (Array.isArray(text) && text.hasOwnProperty('raw')) {
		return text.join("%s");
	}

	return text;
}

function postProcess(text, namedValues, params, count) {

    // test for parsable interval string
    if (/\|/.test(text)) {
    	text = parsePluralInterval(text, count);
    }

    // replace the counter
    if (typeof count == "number") {
    	text = vsprintf(text, [parseInt(count, 10)]);
    }

    // if the text string contains {{Mustache}} patterns we render it as a mini tempalate
    if (/\{\{.*\}\}/.test(text)) {
    	text = Mustache.render(text, namedValues);
    }

    // if we have extra arguments with values to get replaced,
    // an additional substition injects those strings afterwards
    if (/%/.test(text) && params.length) {
    	text = vsprintf(text, params);
    }

    return text;
}

/**
 * splits and parses a phrase for mathematical interval expressions
 */
function parsePluralInterval(phrase, count) {

	var returnPhrase = phrase,
		phrases = phrase.split(/\|/);

	// some() breaks on 1st true
	phrases.some((p) => {

		var [m1, m2] = p.match(/^\s*([\(\)\[\]\d,]+)?\s*(.*)$/);

		// not the same as in combined condition
		if (m1) {

			if (matchInterval(count, m1) === true) {
    			returnPhrase = m2;
    			return true;
			}

		} else {
			returnPhrase = p;
		}
	});

	return returnPhrase;
}

/**
 * test a number to match mathematical interval expressions
 * [0,2] - 0 to 2 (including, matches: 0, 1, 2)
 * ]0,3[ - 0 to 3 (excluding, matches: 1, 2)
 * [1]   - 1 (matches: 1)
 * [20,] - all numbers ≥20 (matches: 20, 21, 22, ...)
 * [,20] - all numbers ≤20 (matches: 20, 21, 22, ...)
 */
function matchInterval(number, interval) {

	interval = ParseInterval(interval);

	if (interval && typeof number == 'number') {

		var { 
			from: {
				included: fromIncluded,
				value:    fromValue
			}, 
			to: {
				included: toIncluded,
				value:    toValue
			}
		} = interval;

		if (fromValue === number) {
			return fromIncluded;
		}

		if (toValue === number) {
			return toIncluded;
		}

		return (
			Math.min(fromValue, number) === fromValue &&
			Math.max(toValue,   number) === toValue
		);
	}

	return false;
}


function translate(locale, singular, plural) {

	var fallback = fallbacks[locale],
		defaultSingular = singular,
		defaultPlural   = plural;

	if (typeof locale == "undefined") {
		locale = defaultLocale;
	}

	if (typeof locales[locale] != "object" && typeof fallback == "string") {
		locale = fallback;
	}

	if (typeof locales[locale] != "object") {
		locale = defaultLocale;
	}

	if (objectNotation) {

    	var indexOfColon = singular.indexOf(':');
    	// We compare against 0 instead of -1 because we don't really expect the string to start with ':'.
    	if (0 < indexOfColon) {
    		defaultSingular = singular.substring(indexOfColon + 1);
    		singular = singular.substring(0, indexOfColon);
    	}

    	if (plural && typeof plural !== 'number') {

      		indexOfColon = plural.indexOf(':');

      		if (0 < indexOfColon) {
		        defaultPlural = plural.substring(indexOfColon + 1);
		        plural        = plural.substring(0, indexOfColon);
			}
		}
	}

	var accessor = localeAccessor(locale, singular),
		mutator  = localeMutator(locale, singular);

	if (plural && !accessor()) {
		mutator({
			'one':   defaultSingular || singular,
			'other': defaultPlural   || plural
		});
	}

	if (!accessor()) {
		mutator(defaultSingular || singular);
	}

	return accessor();
}

/**
 * Allows delayed access to translations nested inside objects.
 * @param {String} locale The locale to use.
 * @param {String} singular The singular term to look up.
 * @param {Boolean} [allowDelayedTraversal=true] Is delayed traversal of the tree allowed?
 * This parameter is used internally. It allows to signal the accessor that
 * a translation was not found in the initial lookup and that an invocation
 * of the accessor may trigger another traversal of the tree.
 * @returns {Function} A function that, when invoked, returns the current value stored
 * in the object at the requested location.
 */
function localeAccessor(locale, singular, allowDelayedTraversal) {

	// Bail out on non-existent locales to defend against internal errors.
	if (typeof locales[locale] != "object") {
		return Function.prototype;
	}

	// Handle object lookup notation
	var indexOfDot = objectNotation && singular.lastIndexOf(objectNotation);

	if (objectNotation && (0 < indexOfDot && indexOfDot < singular.length - 1)) {

		// If delayed traversal wasn't specifically forbidden, it is allowed.
		if (typeof allowDelayedTraversal == "undefined") {
			allowDelayedTraversal = true;
		}

		// The accessor we're trying to find and which we want to return.
		var accessor = null;
		// An accessor that returns null.
		var nullAccessor = () => null;
		// Do we need to re-traverse the tree upon invocation of the accessor?
		var reTraverse = false;

		// Split the provided term and run the callback for each subterm.
		singular.split(objectNotation).reduce((object, index) => {
			// Make the accessor return null.
			accessor = nullAccessor;
			// If our current target object (in the locale tree) doesn't exist or
			// it doesn't have the next subterm as a member...
			if (null === object || !object.hasOwnProperty(index)) {
				// ...remember that we need retraversal (because we didn't find our target).
				reTraverse = allowDelayedTraversal;
				// Return null to avoid deeper iterations.
				return null;
			}
			// We can traverse deeper, so we generate an accessor for this current level.
			accessor = () => object[index];
			// Return a reference to the next deeper level in the locale tree.
			return object[index];

		}, locales[locale]);

		// Return the requested accessor.
		return () =>
			// If we need to re-traverse (because we didn't find our target term)
			// traverse again and return the new result (but don't allow further iterations)
			// or return the previously found accessor if it was already valid.
			( reTraverse ) ? localeAccessor(locale, singular, false)() : accessor()
		;
	} else {
		// No object notation, just return an accessor that performs array lookup.
		return () => locales[locale][singular];
	}
}

/**
 * Allows delayed mutation of a translation nested inside objects.
 * @description Construction of the mutator will attempt to locate the requested term
 * inside the object, but if part of the branch does not exist yet, it will not be
 * created until the mutator is actually invoked. At that point, re-traversal of the
 * tree is performed and missing parts along the branch will be created.
 * @param {String} locale The locale to use.
 * @param {String} singular The singular term to look up.
 * @param {Boolean} [allowBranching=false] Is the mutator allowed to create previously
 * non-existent branches along the requested locale path?
 * @returns {Function} A function that takes one argument. When the function is
 * invoked, the targeted translation term will be set to the given value inside the locale table.
 */
function localeMutator(locale, singular, allowBranching) {

	// Bail out on non-existent locales to defend against internal errors.
	if (typeof locales[locale] != "object") {
		return Function.prototype;
	}

	// Handle object lookup notation
	var indexOfDot = objectNotation && singular.lastIndexOf(objectNotation);

	if (objectNotation && (0 < indexOfDot && indexOfDot < singular.length - 1)) {

		// If branching wasn't specifically allowed, disable it.
		if (typeof allowBranching == "undefined") {
			allowBranching = false;
		}

		// This will become the function we want to return.
		var accessor = null;
		// An accessor that takes one argument and returns null.
		var nullAccessor = () => null;
		// Fix object path.
		var fixObject = () => ({});
		// Are we going to need to re-traverse the tree when the mutator is invoked?
		var reTraverse = false;

		// Split the provided term and run the callback for each subterm.
		singular.split(objectNotation).reduce((object, index) => {
			// Make the mutator do nothing.
			accessor = nullAccessor;
			// If our current target object (in the locale tree) doesn't exist or
			// it doesn't have the next subterm as a member...
			if (null === object || !object.hasOwnProperty(index)) {
				// ...check if we're allowed to create new branches.
				if (allowBranching) {
					// Fix `object` if `object` is not Object.
					if (null === object || typeof object != "object") {
						object = fixObject();
					}
					// If we are allowed to, create a new object along the path.
					object[index] = {};
				} else {
					// If we aren't allowed, remember that we need to re-traverse later on and...
					reTraverse = true;
					// ...return null to make the next iteration bail our early on.
					return null;
				}
			}
			// Generate a mutator for the current level.
			accessor = value => object[index] = value;
			// Generate a fixer for the current level.
			fixObject = () => (object[index] = {});
			// Return a reference to the next deeper level in the locale tree.
			return object[index];

		}, locales[locale]);

		// Return the final mutator.
		return value =>
			// If we need to re-traverse the tree
			// invoke the search again, but allow branching this time (because here the mutator is being invoked)
			// otherwise, just change the value directly.
			( reTraverse ) ? localeMutator(locale, singular, true)(value) : accessor(value)
		;

	} else {
		// No object notation, just return a mutator that performs array lookup and changes the value.
		return value => locales[locale][singular] = value;
	}
}